README.rst
=============================
inforcehub
=============================

.. image:: https://badge.fury.io/py/inforcehub.png
    :target: http://badge.fury.io/py/inforcehub

.. image:: https://travis-ci.org/Inforcehub/inforcehub.png?branch=master
    :target: https://travis-ci.org/Inforcehub/inforcehub

Utilities for data science and customer management


Features
--------

* **colors** module - to enable quick use of inforcehub branded colors
* **anonymize** module - to encrypt customer identifiable data from a dataset and (optionally) create a key


Installation
------------

To install the module in your virtual environment using pip::

    pip install inforcehub

To upgrade your existing inforcehub module installation using pip::

    pip install -U inforcehub
    

Requirements
------------

Python 3.5 or later 


Documentation
-------------

Full documentation can be found at http://inforcehub.rtfd.org



HISTORY.rst
.. :changelog:

History
-------


0.2.2 (2018-11-07)
++++++++++++++++++

* Fixed document build on Read the Docs


0.2.1 (2018-11-02)
++++++++++++++++++

* Fixed dependencies in PyPI


0.2.0 (2018-10-19)
++++++++++++++++++

* Added Anonymize module to encrypt customer identifiable data in a dataset
* Removed PyPy support due to PyPy and numpy incompatibility


0.1.3 (2018-10-16)
++++++++++++++++++

* Minor bug fix


0.1.2 (2018-10-16)
++++++++++++++++++

* Colors module refactored and additional tests written
* Documentation updated
* Added support for Python 3.5+


0.1.1 (2018-10-15)
++++++++++++++++++

* Documentation added


0.1.0 (2018-10-15)
++++++++++++++++++

* First release on PyPI.
* Colors module providing inforcehub brand colors


test/test_colors.py
"""
Tests for color module.
"""
import pytest
from inforcehub import InforcehubColors


class TestInforcehubColors(object):
    @classmethod
    def setup_class(cls):

        cls.c = InforcehubColors()

        cls.core_colors = [
            ("pink", "#CB1B64"),
            ("blue", "#0593CF"),
            ("green", "#079921"),
            ("yellow", "#F3B80C"),
            ("cyan", "#0BD0B1"),
        ]

        cls.neutral_colors = [
            ("mid", "#8C8C8C"),
            ("dark", "#4D4D4D"),
            ("light", "#CCCCCC"),
            ("black", "#011329"),
            ("white", "#FCFCFC"),
        ]

        cls.extra_colors = [
            ("purple", "#5E2099"),
            ("forest", "#137748"),
            ("teal", "#117789"),
            ("violet", "#5446A0"),
            ("orange", "#FC5507"),
            ("red", "#E52019"),
        ]

    @classmethod
    def teardown_class(cls):
        pass

    def test_attributes_return_correct_hex_codes(self):
        """ Tests if all colors return the right hex values """
        for color in self.core_colors:
            assert getattr(self.c, color[0]) == color[1]
        for color in self.neutral_colors:
            assert getattr(self.c, color[0]) == color[1]
        for color in self.extra_colors:
            assert getattr(self.c, color[0]) == color[1]

    def test_returns_valid_hex(self):
        """ Tests if the format is correct """
        for hexcode in self.c.list():
            assert isinstance(hexcode, str)
            assert len(hexcode) == 7
            assert hexcode[0] == "#"
            assert isinstance(int(hexcode[1:], 16), int)  # Checks it is valid hex

    def test_show_method(self):
        """ Tests that the show returns a list """
        assert isinstance(self.c.show(), list)
        assert len(self.c.show()) == 16
        assert self.core_colors[1][0] in self.c.show()
        assert self.neutral_colors[3][0] in self.c.show()
        assert self.extra_colors[0][0] in self.c.show()

    def test_show_method_sublists(self):
        """ Tests that the sublists can be called effectively on the method """
        assert len(self.c.show("all")) == 16
        assert len(self.c.show("colors")) == 11
        assert len(self.c.show("core")) == 5
        assert self.core_colors[4][0] in self.c.show("core")
        assert len(self.c.show("neutral")) == 5
        assert self.neutral_colors[1][0] in self.c.show("neutral")

    def test_show_method_exception(self):
        """ Tests the exceptions are raised for invalid color list """
        with pytest.raises(Exception):
            self.c.show("invalid text")

    def test_list_method(self):
        """ Tests that the list method returns hex value lists correctly """
        assert isinstance(self.c.list(), list)
        assert len(self.c.list()) == 16
        assert self.core_colors[1][1] in self.c.list()
        assert self.neutral_colors[1][1] in self.c.list()
        assert self.extra_colors[1][1] in self.c.list()

    def test_list_method_sublists(self):
        """ Tests that the sublists can be called effectively on the method """
        assert len(self.c.list("all")) == 16
        assert len(self.c.list("colors")) == 11
        assert len(self.c.list("core")) == 5
        assert self.core_colors[3][1] in self.c.list("core")
        assert len(self.c.list("neutral")) == 5
        assert self.neutral_colors[0][1] in self.c.list("neutral")

    def test_list_method_exception(self):
        """ Tests the exceptions are raised for invalid color list """
        with pytest.raises(Exception):
            self.c.list("invalid text")


test/test_anonymize.py
"""
Tests for anonymize module.
"""
import pandas as pd
import pytest
from inforcehub import Anonymize


@pytest.fixture(scope="module")
def test_df():
    # Load test data into dataframe - copy this when running tests
    return pd.read_csv("test/test_fixtures/test_product_data.csv")


class TestInforcehubAnon(object):

    anon = Anonymize(salt="PASSWORDTHING")  # This is our test salt

    def test_anon_init_is_random(self):
        """ Test that new anon instances have randomised salt """
        a = Anonymize()
        b = Anonymize()
        text = "Test thing"
        assert a.salt != b.salt
        assert a.encode(text) != b.encode(text)
        assert a.encode(text) != self.anon.encode(text)

    def test_anon_result_with_known_password(self):
        """ Tests the output hex value is as expected when password is set and known """
        assert self.anon.salt == "PASSWORDTHING".encode()
        assert self.anon.encode("This is a thing") == "9ecdcb0b584d922c1d9af10d77edd522"

    def test_encode_produces_string(self):
        """ Tests the output value of the encode routine """
        assert isinstance(self.anon.encode("Something"), str)
        assert len(self.anon.encode("Something else")) == 32

    def test_raises_exception_on_column_error(self, test_df):
        """ Tests that columns that don't exist raises an exception """
        df = test_df.copy()
        with pytest.raises(Exception):
            self.anon.transform(df, ["First name", "Error name"])
        with pytest.raises(Exception):
            self.anon.transform(df, "Single error column name")

    def test_columns_transformed_OK(self, test_df):
        """ Tests that the test dataframe is converted correctly """
        df = test_df.copy(deep=True)
        self.anon.transform(df, "Contract number")
        assert df.shape == test_df.shape
        assert df["Contract number"].iloc[0] == "c5e8b35adac40efce02cc69487b3addc"
        assert test_df["Contract number"].iloc[0] == 33463634
        assert df["Contract number"].all() != test_df["Contract number"].all()
        assert df["Product"].iloc[2] == "Mousetrap Pro"
        assert df["Product"].all() == test_df["Product"].all()

    def test_pseudo_anon_dataset(self, test_df):
        """ Tests that the pseudo_anonymised dataset is valid """
        df = test_df.copy(deep=True)
        key_df = self.anon.transform(df, ["Contract number", "Status"])
        assert key_df.shape[1] == 4
        assert key_df["Contract number_"].iloc[0] == "c5e8b35adac40efce02cc69487b3addc"
        assert key_df["Contract number"].iloc[0] == 33463634
        assert key_df["Status_"].iloc[3] == "e76e1e08dbbe5b255724b6520cbb5497"
        assert key_df["Status"].iloc[3] == "Paid-up"

    def test_multiple_columns_transformed(self, test_df):
        """ Tests that multiple columns are transformed correctly """
        df = test_df.copy(deep=True)
        self.anon.transform(df, ["First name", "Product", "Start date"])
        assert df.shape == test_df.shape
        for column in ["Contract number", "Last name", "Status", "In-force premium"]:
            assert df[column].all() == test_df[column].all()
        for column in ["First name", "Product", "Start date"]:
            assert df[column].all() != test_df[column].all()
            assert isinstance(df[column].iloc[3], str)
            assert len(df[column].iloc[3]) == 32

    def test_same_names_give_same_hash(self, test_df):
        """ Tests that the same text in the input will have the same hash """
        """ Checks that we haven't updated the salt by accident """
        df = test_df.copy(deep=True)
        self.anon.transform(df, ["Last name", "Product"])
        # Both surnames are Brown
        assert len(df["Last name"].iloc[3]) == 32
        assert df["Last name"].iloc[3] == df["Last name"].iloc[4]
        # Same product names for first and second row
        assert df["Product"].iloc[0] == df["Product"].iloc[1]
        assert df["Product"].iloc[0] != df["Product"].iloc[3]


test/test_inforcehub.py
"""
Tests for `inforcehub` module.
"""


class TestInforcehub(object):
    @classmethod
    def setup_class(cls):
        pass

    @classmethod
    def teardown_class(cls):
        pass

    def test_color_import(self):
        """ Confirm imports are available for modules """
        from inforcehub import InforcehubColors

        c = InforcehubColors()
        c.list()

    def test_anon_import(self):
        """ Confirm imports are available for modules """
        from inforcehub import Anonymize

        anon = Anonymize()
        anon.salt


inforcehub/anonymize.py
import hashlib
import bcrypt
import pandas as pd


class Anonymize:
    """
    A class that will transform a Pandas dataframe into an anonymized
    dataframe using the python hmac package for encryption.

    When instantiating an object of this class, the **salt**
    init attribute can be specified which enables encryption results
    to be reproduced. If none is supplied, a randomized password
    is created instead for security.

    :param: str salt: an optional passphase - if empty, salt will be randomized
    """

    def __init__(self, salt=None):
        """"""
        if salt:
            # Convert to bytestring
            self.salt = str(salt).encode()
        else:
            self.salt = bcrypt.gensalt()

    def encode(self, text):
        """
        Provides the encryption of a single number, text or date

        :param: text: the text, value or date to be encrypted

        :returns: the hexidecimal encrypted value
        :rtype: str
        """
        return hashlib.md5(self.salt + str(text).encode()).hexdigest()

    def transform(self, df, columns, verbose=False):
        """
        Encrypts the selected columns in a dataframe and returns
        an anonymized dataframe.

        If required, also returns a dataframe of column pairs showing
        the encrypted and original data to be used as a
        pseudo-anonymization key.

        :param: pd.DataFrame() df: A Pandas dataframe to be transformed
        :param: list columns: A list of columns to be transformed
        :param: str verbose: (default=False) If true will print status

        :returns: a pseudo-anonymization lookup table
        :rtype: pd.DataFrame()
        """
        # In case user puts in single column name as a string not a list
        if isinstance(columns, str):
            columns = [columns]

        if verbose:
            print("Will convert columns: %s" % ", ".join(columns))
            print("Encrypting %i rows per column ...\n" % df.shape[0])

        lookup = pd.DataFrame()
        for column in columns:

            if column not in df.columns:
                raise Exception("Column %s cannot be found in dataframe" % column)

            lookup[column] = df[column]

            df[column] = df[column].map(lambda x: self.encode(x))

            lookup[column + "_"] = df[column]

            if verbose:
                print("Finished encrypting column %s" % column)

        return lookup


inforcehub/colors.py
class InforcehubColors:
    """
    A class that will return the inforcehub brand colors as single hex codes
    or lists for use in various packages such as matplotlib.

    The objective of this package is to make it easier and quicker to create
    consistent-looking charts using branded colors
    """

    # Core colors
    pink = "#CB1B64"
    blue = "#0593CF"
    green = "#079921"
    yellow = "#F3B80C"
    cyan = "#0BD0B1"
    core = ["pink", "blue", "green", "yellow", "cyan"]

    # Neutral colors
    mid = "#8C8C8C"
    dark = "#4D4D4D"
    light = "#CCCCCC"
    black = "#011329"
    white = "#FCFCFC"
    neutral = ["mid", "dark", "light", "black", "white"]

    # Extra colors (not core but useful when we need more colors)
    forest = "#137748"
    purple = "#5E2099"
    teal = "#117789"
    violet = "#5446A0"
    orange = "#FC5507"
    red = "#E52019"
    extra = ["forest", "purple", "teal", "violet", "orange", "red"]

    colors = core + extra
    all = core + extra + neutral

    @classmethod
    def show(cls, sublist="all"):
        """
        Returns a list of the color names

        :param: str sublist: (default='all') options 'colors'/'core'/'neutral'/'all'

        :returns: a list of color names
        :rtype: list

        """
        if sublist is not None:
            if sublist not in ["core", "neutral", "colors", "all"]:
                raise Exception(
                    'List type can only be "core", "neutral", "colors", or "all"'
                )

        list_ = []
        for color in getattr(cls, sublist):
            list_.append(color)
        return list_

    @classmethod
    def list(cls, sublist="all"):
        """
        Returns a list of the color hex codes for use in Matplotlib
        or other tools which want a set of colors to choose from

        :param: str sublist: (default='all') options 'colors'/'core'/'neutral'/'all'

        :returns: a list of color hex codes
        :rtype: list

        """
        if sublist is not None:
            if sublist not in ["core", "neutral", "colors", "all"]:
                raise Exception(
                    'List type can only be "core", "neutral", "colors", or "all"'
                )

        list_ = []
        for color in getattr(cls, sublist):
            list_.append(getattr(cls, color))
        return list_
